skimage/measure/_marching_cubes_lewiner_cy.pyx

```diff
@@ -929,13 +929,15 @@ cdef class LutProvider:
 
         self.SUBCONFIG13 = Lut(SUBCONFIG13)
 
-
-def marching_cubes(float [:, :, :] im not None, double isovalue, LutProvider luts, int st=1, int classic=0):
+def marching_cubes(float [:, :, :] im not None, double isovalue, LutProvider luts, int st=1, int classic=0,
+                   int [:, :, :] mask=None):
```

Comment:
If you declare `mask` as `bint`, you can avoid its conversion to `int32` in *_marching_cube_lewiner.py* line 153 (see this [ref](https://github.com/cython/cython/wiki/DifferencesFromPyrex) for example).

Reply:
Thank you very much for the advice! It is the first time I wrote something on cython, I remember to be stuck on this point for a while and I was only able to make it run with the hack. I just pushed the suggested changes

Reply:
Oops, looks like I misled you! Sorry...
Looking at one of my previous PR, I realized that a better way is:
```python
def marching_cubes(float [:, :, :] im not None, double isovalue, LutProvider luts, int st=1, int classic=0,
                   np.ndarray[np.npy_bool, ndim=3, cast=True] mask=None):
```

Reply:
and that compiles in C? I remember that at some point I had it some python comparison and it slowed down the whole thing by a lot

Reply:
Cython's docs recommend [memoryviews](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html) over the old `np.array` API. 

I haven't looked through the rest of the PR but if `mask` is a boolean array you can use it without conversion by constructing a view `mask.view(np.uint8)` beforehand which is a compatible dtype to `bint[:, :, :]` (or `np.uint8_t[:, :, :]`, same thing memorywise!). Cython's memoryview API can't yet deal with boolean arrays automatically. That should be the most efficient way.

If you need even faster indexing and your input arrays are contiguous take a look at [this section](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#c-and-fortran-contiguous-memoryviews).

Reply:
Hi @lagru, thank you for this ref...
But the main advantage of the NumPy array buffer syntax here is that it explicitly deal with data cast to bool.
Cython MemoryView doesn't give much performance here since we do not enforce `mask` to be memory contiguous.

Reply:
@Leguark, yes this is the numpy array buffer syntax, supported by Cython ;-)

Reply:
> But the main advantage of the NumPy array buffer syntax here is that it explicitly deal with data cast to bool.

Note: using cast will be no longer necessary once we upgrade our dependency to Cython 0.29.14 wich contains a [fix added support for bool](https://github.com/cython/cython/pull/2676). 

> Cython MemoryView doesn't give much performance here since we do not enforce mask to be memory contiguous.

Have you benchmarked that? 

I'd guess that `mask` is a contiguous array most of the time so we could enforce it? Using `np.asarray(mask, dtype=bool, order="C").view(np.uint8)` together with the memoryview API might yield noticeable improvements for the common case considering that we check that array deep inside 3 nested loops.

Otherwise I'd still prefer not to mix both APIs here and go with the recommended one. Although I grant you that in this case the old API has an advantage over the new one and I won't complain if there is no performance difference. :wink:

Reply:
I did not benchmarked, just guessing :-p

> I'd guess that `mask` is a contiguous array most of the time so we could enforce it? Using `np.asarray(mask, dtype=bool, order="C").view(np.uint8)` together with the memoryview API might yield noticeable improvements for the common case considering that we check that array deep inside 3 nested loops.

Good suggestion! :wink: 

Both options are OK for me, so @Leguark, its better to implement @lagru suggestion, and sorry once again that my proposition gave you extra work.